deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/BasicGradientsAccumulator.java

```diff
@@ -0,0 +1,259 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This class provides accumulation for gradients for both input (i.e. updates coming from network) and output (comint from one ore more models training at the same time)
+ *
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class BasicGradientsAccumulator implements GradientsAccumulator {
+
+    protected MessageHandler handler;
+
+    // here we'll store messages coming from "somewhere else"
+    protected transient Queue<INDArray> gradients;
+
+    // this field stores current accumulated
+    protected transient INDArray storage;
+
+    // this field stores updates ready-to-apply
+    protected transient INDArray updates;
+
+    // this counter tracks number of messages generated by this accumulation
+    protected transient AtomicLong ownCounter = new AtomicLong(0);
+
+    // this counter tracks number of messages received from somewhere
+    protected transient AtomicLong extCounter = new AtomicLong(0);
+
+    // FIXME: this mechanics should be improved i think.
+    protected int[] shape;
+    protected char ordering;
+
+    protected int parties = 0;
+    protected CyclicBarrier barrier;
+    protected AtomicLong firstOne = new AtomicLong(-1L);
+    protected List<INDArray> candidates = new CopyOnWriteArrayList<>();
+
+    protected ReentrantReadWriteLock updatesLock = new ReentrantReadWriteLock();
+    protected AtomicBoolean hasSomething = new AtomicBoolean(false);
+
+    /**
+     * Creates new GradientsAccumulator with starting threshold of 1e-3
+     */
+    public BasicGradientsAccumulator(int parties) {
+        this(parties, new LocalHandler());
+    }
+
+    /**
+     * Creates new GradientsAccumulator with custom starting threshold
+     *
+     * @param handler MessageHandler instance that'll be used for communication purposes
+     */
+    public BasicGradientsAccumulator(int parties, @NonNull MessageHandler handler) {
+        this.gradients = new LinkedTransferQueue<>();
+        this.handler = handler;
+
+        this.handler.initialize(this);
+        this.parties = parties;
+        barrier = new CyclicBarrier(parties);
+    }
+
+    /**
+     * This method applies accumulated updates via given StepFunction
+     *
+     * @param function
+     * @param params
+     */
+    @Override
+    public void applyUpdate(StepFunction function, INDArray params) {
+
+        try {
+            updatesLock.readLock().lock();
+
+            firstOne.compareAndSet(-1L, Thread.currentThread().getId());
+
+            if (hasSomething.get())
+                function.step(params, updates);
+
+            barrier.await();
+            if (firstOne.get() == Thread.currentThread().getId()) {
+                // one thread just nullifies this array
+                updates.assign(0.0);
+                hasSomething.set(false);
+
+                firstOne.set(-1L);
+            }
+
+            updatesLock.readLock().unlock();
```

Comment:
What happens if anything in this try block throws an exception before .unlock()?
Same thing for other uses of locks/barriers in this class.
It's pretty unlikely, but not impossible.